soco/zonegroupstate.py

```diff
@@ -148,8 +150,88 @@ def poll(self, soco):
+        except SoCoUPnPException as soco_upnp_exception:
+            _LOG.debug(
+                "Exception (%s) raised on 'GetZoneGroupState()'",
+                soco_upnp_exception,
+            )
+
+            if config.ZGT_EVENT_FALLBACK is False:
+                _LOG.debug("ZGT event fallback disabled (config.ZGT_EVENT_FALLBACK)")
+                raise NotSupportedException(
+                    "'GetZoneGroupState()' call fails on large Sonos systems "
+                    "and event fallback is disabled"
+                ) from soco_upnp_exception
+
+            _LOG.debug("Falling back to using a ZGT event")
+            try:
+                self.update_zgs_by_event(soco)
+            except Exception as soco_exception:
+                raise soco_exception from soco_upnp_exception
+
+    def update_zgs_by_event(self, speaker):
+        """
+        Fall back to updating the ZGS using a ZGT event.
+        Use of the 'events_twisted' module is not currently supported.
+        """
+        if config.EVENTS_MODULE.__name__ == "soco.events":
+            _LOG.debug("Updating ZGS using standard 'events' module")
+            self.update_zgs_by_event_default(speaker)
+
+        if config.EVENTS_MODULE.__name__ == "soco.events_asyncio":
+            _LOG.debug("Updating ZGS using 'events_asyncio' module")
+            # Explicit asyncio event loop control required for Python 3.6
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
+            loop.run_until_complete(ZoneGroupState.update_zgs_by_event_asyncio(speaker))
+            asyncio.set_event_loop(None)
+            loop.close()
+            # From Python 3.7, we can just use the single statement:
+            # asyncio.run(ZoneGroupState.update_zgs_events_asyncio(speaker))
+
+        if config.EVENTS_MODULE.__name__ == "soco.events_twisted":
+            raise SoCoException(
+                "ZGT event fallback not supported when using the "
+                "'events_twisted' module"
+            )
+
+    def update_zgs_by_event_default(self, speaker):
+        """
+        Update the ZGS using the default events module.
+        """
+        sub = speaker.zoneGroupTopology.subscribe()
+        event = sub.events.get(timeout=1.0)
+        sub.unsubscribe()
+        zgs = event.variables.get("zone_group_state")
+        self.process_payload(payload=zgs, source="event", source_ip=speaker.ip_address)
+
+    @staticmethod
+    async def update_zgs_by_event_asyncio(speaker):
+        """
+        Update ZGS using events_asyncio. When the event is received,
+        the events_asyncio notify handler will call 'process_payload' with
+        the updated ZGS.
+        """
+        from . import events_asyncio  # pylint: disable=C0415
+
+        event_listener_is_running = events_asyncio.event_listener.is_running
+        sub = await speaker.zoneGroupTopology.subscribe()
+        await asyncio.sleep(1.0)
```

Comment:
I'm wondering of the "sleep 1" is necessary, when the other variants only have a timeout? But also I don't really understand the rest. Forgive me, async is not me strongest suit. But shouldn't this somehow get and process and events like the non async version and wont the awaiting of that event remove remove the need for the the sleep and be equivalent to the timeout?

Reply:
Not my strongest area, either. I think a `sleep()` is needed to give other coroutines a chance to run. However, I'll experiment with shorter durations, and also with omitting it completely.

Reply:
But what about the question about actually getting the payload. It is somehow happening behind the scenes or am I missing it?

Reply:
> Not my strongest area, either. I think a sleep() is needed to give other coroutines a chance to run. However, I'll experiment with shorter durations, and also with omitting it completely.

I think is the sort of thing that async people use Futures or Promises or whatever else they are called in difference asyncio frameworks for. Wrapping a corutine in an awaitable objects which "returns" when the coroutine have been executed. But I'm basically just saying words here.

Reply:
> But what about the question about actually getting the payload. It is somehow happening behind the scenes or am I missing it?

Yes, it's automatically handled in `events_asyncio` when the event notification is received: https://github.com/SoCo/SoCo/blob/f41b6e81b08ff6d615a1ae3a034c2dbb3be81cae/soco/events_asyncio.py#L140-L147

Reply:
> I think a `sleep()` is needed to give other coroutines a chance to run. However, I'll experiment with shorter durations, and also with omitting it completely.

I think I understand what's happening now:
- The `asyncio.sleep()` is needed to introduce enough delay between `subscribe()` and `unsubscribe()` to allow an event to be generated by the Sonos player and received by SoCo, and it allows the event handler coroutine to run
- On my system I'm seeing an event response within ~25ms of the `subscribe()`
- Hence, I'm going to set the delay at 250ms; we can adjust this later (or make it configurable) if it turns out empirically to be too short